agent/php_stacked_segment.h

```diff
@@ -155,6 +164,386 @@
+ *                               |                           |
+ *                               *B                          *B <
+ *                               |                           |
+ *                               *C <                        c
+ *
+ * nr_php_observer_fcall_begin(C) starts *C gets started as child
+ * of *B. Function C throws an exception which B catches but 
+ * nr_php_observer_fcall_end(C) is not called so *C remains the 
+ * current segment. B catches the exception and throws another 
+ * exception which triggers the exception hook. At this point we realize the current
+ * exception->This value indicates another function is active.  Because we
+ * received no nr_php_observer_fcall_end up to that point, we know the exception
+ * was uncaught until the exception->This function. We check the global
+ * exception hook and see it has a value and that the global
+ * uncaught_exception_this also matches the current segment `this`. Time to
+ * apply the exception and clean up dangling segments. We pop the current
+ * segment *C and apply the exception. Because it has an exception, the segment
+ * is kept so we copy the contents of the stacked segment *C into a segment c we
+ * obtained from the slab allocator, and we make c a child of the stacked
+ * segment *B which becomes the current segment.
+ *
+ *  root
+ *   |
+ *   *A
+ *    |
+ *    *B <
+ *    |
+ *    c
+ *
+ * current stacked_segment->metadata->this now equals the exception->This. so we
+ * reserve judgement on what eventually happens to segment *B and *B becomes the
+ * current segment with the new active exception stored. Any subsequent dangling
+ * segments are cleaned when the next scenario 1-5 occurs.
+ *
+ * case 5 in newrelic APIs that depend on having the current segment
+ *  root <                      root                      root
+ *                               |                         |
+ *                               *A <                      *A
+ *                                                         |
+ *                                                         *B
+ *
+ * We start out with a root segment, OAPI calls nr_php_observer_fcall_begin(A),
+ * and it starts stacked segment *A and then nr_php_observer_fcall_begin(B)
+ * starts stacked segment *B as a child of *A.
+ *
+ *  root                       root
+ *   |                           |
+ *   *A                          *A
+ *   |                           |
+ *   *B                          *B <
+ *   |                           |
+ *   *C <                        c
+ *
+ * nr_php_observer_fcall_begin(C) starts *C gets started as child
+ * of *B. Function C throws an uncaught exception which B does not catch so
+ * neither nr_php_observer_fcall_end(B) nor nr_php_observer_fcall_end(C) is
+ * called and *C remains the current segment. A catches the exception and calls
+ * newrelic_notice_error. We check the `this` value of the function that called
```

Comment:
The `this` value check happens at the beginning of the `newrelic_notice_error`, yes? If so then consider making it clear:
```suggestion
 * newrelic_notice_error. In newrelic_notice_error we check the `this` value of the function that called
```

Reply:
None of the APIs do this checking explicitly.  The API functions call `nr_php_api_ensure_current_segment` (before doing anything related to segments).  `nr_php_api_ensure_current_segment` passes the `This` pointer of the calling function to `nr_php_observer_handle_uncaught_exception` which calls [nr_php_observer_handle_uncaught_exception](https://github.com/newrelic/newrelic-php-agent/pull/580/files#diff-beedad7bbfc9687348e557b4752fbf6f54e4407a6e3177e8cabaaec7e3ffd625R1864) which does the checking in php_execute.c.  `nr_php_observer_handle_uncaught_exception` is the function that all    code paths will call when they detect dangling segments due to exceptions.  Please see clarification here: [80f1acd](https://github.com/newrelic/newrelic-php-agent/pull/580/commits/80f1acd0406ea78a52ac0bb0bf8ebc604bad9903)